### PR TITLE
[IOTDB-1597] Finish DIFF encoding algorithm which is reserved in enum before

### PR DESCRIPTION
**Pull request by 5Reasons**
In the enum class: TSEncoding, there's  DIFF encoding, so when user want to choose a type of encoding algorithm(or input the wrong encoding type name), DIFF encoding is in the output hint and is one of IoTDB's recommending encoding type, 
**but in IoTDB, DIFF encoding has not been added before,** which I think may cause some problems.
I think it's necessary to finish it.

Both DIFF and TS_2DIFF encoding are based on delta encoding algorithm, 
so I finish DIFF using similar interfaces and structures like TS_2DIFF to make it easier to understand.

### CODE REVIEW

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+    protected static final int BLOCK_DEFAULT_SIZE = 128;
+    private static final Logger logger = LoggerFactory.getLogger(DeltaBinaryEncoder.class);
+    protected ByteArrayOutputStream out;
+    protected int blockSize;
+    // input value is stored in deltaBlackBuffer temporarily
+    protected byte[] encodingBlockBuffer;
+
+    protected int writeIndex = -1;//defalut value?
+    protected int writeWidth = 0;
```
Comment (HTHou) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java:59]: The indent of this java file is not correct. Please fix it.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+    }
+
+    private void writeHeaderToBytes() throws IOException {
+        ReadWriteIOUtils.write(writeIndex, out);
+        ReadWriteIOUtils.write(writeWidth, out);
+        writeHeader();
+    }
+
+
+    //这其实是个写入函数，在真正flush清空之前，把东西都写到out里
```
Comment (HTHou) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java:97]: Better to use English.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        writeDataWithMinWidth();
+
+        reset();
+        writeIndex = -1;
+    }
+
+    /**
+     * calling this method to flush all values which haven't encoded to result byte
+     * array.
+     * 调用flushBlockBuffer函数
```
Comment (HTHou) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java:116]: Same.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        /**
+         * constructor of IntDeltaEncoder which is a sub-class of DeltaBinaryEncoder.
+         *
+         * @param size - the number how many numbers to be packed into a block.
+         */
+        public IntDeltaEncoder(int size) {
+            super(size);// choose TS_2DIFF + initialize deltaSize=size
+            deltaBlockBuffer = new int[this.blockSize]; //deltaBlockBuffer: 暂时性存储待压缩的数据
+            encodingBlockBuffer = new byte[blockSize * 4]; //目前的理解：压缩后存放的地方
+            reset();//初始化
```
Comment (HTHou) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java:147]: Same

File: tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DiffEncoderIntegerTest.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Random;
+
+import static org.junit.Assert.assertEquals;
+
+public class DiffEncoderIntegerTest {
+
+  private static final int ROW_NUM = 100_000_00;//行数，数组的大小
```
Comment (HTHou) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DiffEncoderIntegerTest.java:35]: Use English

File: tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DiffEncoderIntegerTest.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    }
+    shouldReadAndWrite(data, ROW_NUM);
+  }
+
+  @Test
+  public void testRandom() throws IOException {
+    int data[] = new int[ROW_NUM];
+    for (int i = 0; i < ROW_NUM; i++) {
+      data[i] = ran.nextInt();
+      System.out.printf("%d ",data[i]);
```
Comment (HTHou) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DiffEncoderIntegerTest.java:77]: This line will print too mush lines...

**Review by HTHou — CHANGES_REQUESTED** [review]

File: tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DiffEncoderIntegerTest.java (hunk 6 above)
Comment (5Reasons) [reply tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DiffEncoderIntegerTest.java:77]: Sorry for not finding the problems before submitting. I will fix them.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+/**
+ * <p>
+ * DiffEncoder is a encoder for compressing data in type of INT32(integer) and
+ * INT64(long). It is based in delta-encoding arithmetic.We adapt a hypothesis
```
Comment (kr11) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java:34]: arithmetic.We adapt -> arithmetic. We adapt 
space

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+ * that contiguous data points have similar values.Thus the difference value of
+ * two adjacent points is smaller than those two point values. One integer in
+ * java takes 32-bits. If a positive number is less than 2^m, the bits of this
+ * integer which index from m to 31 are all 0. Given an array which length is n,
+ * if all values in input data array are all positive and less than 2^m, we need
+ * actually m*n, but not 32*n bits to store the array.
+ * </p>
+ * <p>
+ * DiffEncoder calculates difference between two adjacent. Then it saves the delta
+ * value. Then it statistics the longest bit length {@code m} it takes for each
```
Comment (kr11) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java:44]: maybe "values" better

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+ * delta value, which means the bit length that maximum delta value takes. Only
+ * the low m bits are saved into result byte array for all delta values.
```
Comment (kr11) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java:46]: m -> {@code m}
It's better to be consistent with the above.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+ * </p>
+ */
+public abstract class DiffEncoder extends Encoder {
+
+    protected static final int BLOCK_DEFAULT_SIZE = 128;
+    private static final Logger logger = LoggerFactory.getLogger(DeltaBinaryEncoder.class);
```
Comment (kr11) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java]: The logger should be initialized by `DiffEncoder`, not another class.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+    protected ByteArrayOutputStream out;
+    protected int blockSize;
+    // input value is stored in deltaBlackBuffer temporarily
+    protected byte[] encodingBlockBuffer;
+
+    protected int writeIndex = -1;//defalut value?
```
Comment (kr11) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java]: What's the meaning of this annotation? We should explain something, not ask :)

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+    protected int writeWidth = 0;
+
+    /**
+     * constructor of DeltaBinaryEncoder.
```
Comment (kr11) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java]: All `DeltaBinaryEncoder` should be removed carefully before it's passed.

Comment (5Reasons) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/DiffEncoder.java]: I have fixed them : ) and I will pay more attention on annotation errors in the future

File: docs/UserGuide/Operation Manual/DML Data Manipulation Language.md [hunk 13]
```diff
@@ -414,10 +416,11 @@ Example 1: get the last point of root.ln.wf01.wt01.speed:
 |  5   | root.ln.wf01.wt01.speed | 100   |
 ```
 
-Example 2: get the last speed, status and temperature points of root.ln.wf01.wt01
+Example 2: get the last speed, status and temperature points of root.ln.wf01.wt01,
+whose timestamp in range [5, 10).
 
 ```
-> select last speed, status, temperature from root.ln.wf01.wt01
```
Comment (jixuan1989) [inline docs/UserGuide/Operation Manual/DML Data Manipulation Language.md:420]: Hm.. I think the changes here have no relation to this PR.

Comment (5Reasons) [reply docs/UserGuide/Operation Manual/DML Data Manipulation Language.md]: Yes..that may caused by my misoperation on Git. I will roll back the version, submit again.

**Comment by jixuan1989** [comment]
please apply this patch to solve the conflicts.
`git apply solve_conflict.txt`
[solve_conflict.txt](https://github.com/apache/iotdb/files/5478979/solve_conflict.txt)

**Comment by jixuan1989** [comment]
1.  there are too many duplications between `DIFFEncoder` and `DeltaBinaryEncoder`. Better to inherit `DeltaBinaryEncoder`, rather than copy codes.

2. What is the benefit to add such a method? Trying to give some experimental results to show when should we use it...

**Comment by 5Reasons** [comment]
> 1. there are too many duplications between `DIFFEncoder` and `DeltaBinaryEncoder`. Better to inherit `DeltaBinaryEncoder`, rather than copy codes.
> 2. What is the benefit to add such a method? Trying to give some experimental results to show when should we use it...

I have made some experiment abouot it(https://cwiki.apache.org/confluence/display/IOTDB/DIFF+encoding%27s+perfomance).
As I said in email-list, this method's performance is not better than TS_2DIFF (included in 'DeltaBinaryEncoder'), only when the data set is in best circumstances, the new method can save very little room (about 4 bytes, in head)

I add it beacuse DIFF is a reserved method, it's one of the recommending method when user want to choose a encoding method in IoTDB, but it's not finished before. It's in the system prompt, but user can't chooose it
system_prompt

**Comment by JackieTien97** [comment]
Hi, please resolve the conflicts

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/decoder/Decoder.java [hunk 14]
```diff
@@ -112,6 +112,16 @@ public static Decoder getDecoderByType(TSEncoding encoding, TSDataType dataType)
         }
       case DICTIONARY:
         return new DictionaryDecoder();
+      case TS_2DIFF:
```
Comment (HTHou) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/decoder/Decoder.java:115]:
```suggestion
      case DIFF:
```

File: tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java [hunk 15]
```diff
@@ -96,15 +96,24 @@ private void writeData(int[] data, int length) {
   }
 
   private void shouldReadAndWrite(int[] data, int length) throws IOException {
-    // System.out.println("source data size:" + 4 * length + " byte");
+    System.out.println("source data size:" + 4 * length + " byte");
```
Comment (HTHou) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java:99]:
```suggestion
```

File: tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java [hunk 16]
```diff
@@ -96,15 +96,24 @@ private void writeData(int[] data, int length) {
...
     out = new ByteArrayOutputStream();
+
+    long encodeStart=System.nanoTime();
```
Comment (HTHou) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java:102]:
```suggestion
```

File: tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java [hunk 17]
```diff
@@ -96,15 +96,24 @@ private void writeData(int[] data, int length) {
...
     writeData(data, length);
+    Long encodeEnd=System.nanoTime();
+    System.out.println("encode take(ns): "+(encodeEnd-encodeStart));
+
```
Comment (HTHou) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java:106]:
```suggestion
```

File: tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java [hunk 18]
```diff
@@ -96,15 +96,24 @@ private void writeData(int[] data, int length) {
...
     byte[] page = out.toByteArray();
-    // System.out.println("encoding data size:" + page.length + " byte");
+    System.out.println("encoding data size:" + page.length + " byte");
```
Comment (HTHou) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java:108]:
```suggestion
```

File: tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java [hunk 19]
```diff
@@ -96,15 +96,24 @@ private void writeData(int[] data, int length) {
...
     buffer = ByteBuffer.wrap(page);
     int i = 0;
+
+    Long decodeStart=System.nanoTime();
```
Comment (HTHou) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java:112]:
```suggestion
```

File: tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java [hunk 20]
```diff
@@ -96,15 +96,24 @@ private void writeData(int[] data, int length) {
...
     while (reader.hasNext(buffer)) {
       assertEquals(data[i++], reader.readInt(buffer));
     }
+    Long decodeEnd=System.nanoTime();
+    System.out.println("decode take(ns): "+(decodeEnd-decodeStart));
```
Comment (HTHou) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/encoding/decoder/delta/DeltaBinaryEncoderIntegerTest.java:117]:
```suggestion
```

**Review by HTHou — APPROVED** [review]
Looks better now.

**Pull request closed without merging**